src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -657,55 +658,67 @@ internal static int GetId(RuntimeTypeHandle typeHandle)
             // check whether a corresponding update is required in ClassDataContract.IsNonAttributedTypeValidForSerialization
             private static DataContract CreateDataContract(int id, RuntimeTypeHandle typeHandle, Type type)
             {
-                lock (s_createDataContractLock)
+                DataContract dataContract = s_dataContractCache[id];
```

Comment:
> There was an iteration of this fix which uses locks more properly but that introduces performance regression in common scenarios.

That's why ```ConcurrentDictionary<...>``` exists.

> The full framework is using Dictionary<...> for this fix

In that case it appears to be flawed in this regard as well, unless there's something else I'm missing that's making it safe.